src/coreclr/jit/importer.cpp

```diff
@@ -12429,12 +12429,29 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
             // Jit can only inline expand the normal CHKCASTCLASS helper.
             canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS);
         }
-        else
+        else if (helper == CORINFO_HELP_ISINSTANCEOFCLASS)
+        {
+            // If the class is exact, the jit can expand the IsInst check inline.
+            canExpandInline = isClassExact;
+        }
+
+        if (isClassExact && ((helper == CORINFO_HELP_ISINSTANCEOFARRAY) || (helper == CORINFO_HELP_CHKCASTARRAY)))
         {
-            if (helper == CORINFO_HELP_ISINSTANCEOFCLASS)
+            // We have "obj isinst/castclass T[]" and in case if T is sealed we can still perform
+            // the fast pMT check (in case if T[] -> obj.GetType is legal in jit-time)
+            bool                 isExact = false;
+            bool                 nonNull = false;
+            CORINFO_CLASS_HANDLE op1Cls  = gtGetClassHandle(op1, &isExact, &nonNull);
```

Comment:
Why do you need to be looking at the op1 class handle at all to decide whether it is ok to inline the check?

Reply:
Good point, started with a too conservative version, let's see if it works without it 🙂